### PR TITLE
String fixes

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: spec/ruby/core/string/index_spec.rb [hunk 1]
```diff
@@ -260,6 +260,7 @@
     "blablabla".index(/.{10}/).should == nil
     "blaxbla".index(/.x/, 3).should == nil
     "blaxbla".index(/..x/, 2).should == nil
+    "I’ve got a multibyte character.\n".index "\n\n"
```
Comment (bjfish) [inline spec/ruby/core/string/index_spec.rb]: Is there an expectation or matcher for this? https://github.com/ruby/spec/blob/master/CONTRIBUTING.md#matchers-and-expectations

Comment (eregon) [reply spec/ruby/core/string/index_spec.rb]: It's missing an expectation entirely here.

File: spec/ruby/library/yaml/load_spec.rb [hunk 2]
```diff
@@ -34,6 +34,27 @@
...
+      [240, 159, 140, 181],
+      # "--- 🌵 and some text" => "🌵 and some text"
+      [45, 45, 45, 32, 240, 159, 140, 181, 32, 97, 110, 100, 32, 115, 111, 109, 101, 32, 116, 101, 120, 116] =>
+      [240, 159, 140, 181, 32, 97, 110, 100, 32, 115, 111, 109, 101, 32, 116, 101, 120, 116],
+      # "--- Some text 🌵 and some text" => "Some text 🌵 and some text"
+      [45, 45, 45, 32, 83, 111, 109, 101, 32, 116, 101, 120, 116, 32, 240, 159, 140, 181, 32, 97, 110, 100, 32, 115, 111, 109, 101, 32, 116, 101, 120, 116] =>
+      [83, 111, 109, 101, 32, 116, 101, 120, 116, 32, 240, 159, 140, 181, 32, 97, 110, 100, 32, 115, 111, 109, 101, 32, 116, 101, 120, 116]
+    }
+    yaml_and_strings.each do |yaml, str|
+      YAML.load(yaml.pack("c*").force_encoding("UTF-8")).should == str.pack("c*").force_encoding("UTF-8")
```
Comment (eregon) [inline spec/ruby/library/yaml/load_spec.rb]: This should be `pack("C*")`, isn't it?

File: spec/ruby/core/string/index_spec.rb (hunk 1 above)
Comment (aardvark179) [reply spec/ruby/core/string/index_spec.rb]: Oops. The failure was that it looped forever, so I hadn't noticed I'd missed the matcher entirely. :grin:

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 3]
```diff
@@ -796,12 +796,12 @@ private int countSlow(DynamicObject string, DynamicObject... otherStrings) {
         public abstract DynamicObject executeDeleteBang(VirtualFrame frame, DynamicObject string, Object[] args);
 
         @Specialization(guards = "isEmpty(string)")
-        public DynamicObject deleteBangEmpty(DynamicObject string, Object[] args) {
+        public DynamicObject deleteBangEmpty(VirtualFrame frame, DynamicObject string, Object[] args) {
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: Wasn't the issue here the return type with `deleteBang`? I'd rather not add the `VirtualFrame` argument if we're not using it.

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 4]
```diff
@@ -3499,44 +3488,29 @@ public Object stringCharacterIndex(DynamicObject string, DynamicObject pattern,
...
-                } else {
+                if (matchesAtOffset(p, e, pp, pe, stringBytes, patternBytes)) {
                     return index;
                 }
             }
 
             return nil();
         }
 
+        private static boolean matchesAtOffset(int offset, int limit, int patternOffset, int patternLimit, byte[] stringBytes, byte[] patternBytes) {
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: Other than having different limits, this looks quite similar to `ArrayUtils.memcmp`. Maybe we just need a new form of that call?

**Comment by eregon** [comment]
@nirvdrum Can you re-review?

**Pull request merged**